### PR TITLE
feat(anta.tests)!: Adding the test case to verify BGP neighbor routes

### LINKED ISSUE #482: Add the test case to verify the BGP redistribution

### PR DESCRIPTION
**Pull request by MaheshGSLAB**
# Description

Add the test case to verify the BGP neighbor routes 

Run the testcase by adding catalog as below:
```
anta.tests.routing:
  bgp:
    - VerifyBGPExchangedRoutes:
        bgp_peers:
          - peer_address: 172.30.255.5
            vrf: default
            advertised_routes:
              - 192.0.254.5/32
            received_routes:
              - 192.0.255.4/32
          - peer_address: 172.30.255.1
            vrf: default
            advertised_routes:
              - 192.0.255.1/32
              - 192.0.254.5/32
            received_routes:
              - 192.0.254.3/32
```
1. Testcase pass when routes are correct:
```
                                                                                          All tests results                                                                                           
┏━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Device    ┃ Test Name                ┃ Test Status ┃ Message(s) ┃ Test description                                                                                                 ┃ Test category ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ s1-spine1 │ VerifyBGPExchangedRoutes │ success     │            │ Verifies if BGP peers have correctly advertised/received routes with type as valid and active for a specified    │ routing, bgp  │
│           │                          │             │            │ VRF.                                                                                                             │               │
└───────────┴──────────────────────────┴─────────────┴────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────┘
```
2. Testcase fail when routes are missing:
```
                                                                                          All tests results                                                                                           
┏━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Device    ┃ Test Name                ┃ Test Status ┃ Message(s)                                                    ┃ Test description                                              ┃ Test category ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ s1-spine1 │ VerifyBGPExchangedRoutes │ failure     │ Following BGP peers are not found or routes are not exchanged │ Verifies if BGP peers have correctly advertised/received      │ routing, bgp  │
│           │                          │             │ properly:                                                     │ routes with type as valid and active for a specified VRF.     │               │
│           │                          │             │ {'bgp_peers': {'172.30.11.1': {'MGMT': 'Not configured'}}}    │                                                               │               │
└───────────┴──────────────────────────┴─────────────┴───────────────────────────────────────────────────────────────┴───────────────────────────────────────────────────────────────┴───────────────┘
``` 

3. Testcase fail when routes are invalid or inactive
```
 
                                                                                          All tests results                                                                                           
┏━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Device    ┃ Test Name                ┃ Test Status ┃ Message(s)                                                    ┃ Test description                                              ┃ Test category ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ s1-spine1 │ VerifyBGPExchangedRoutes │ failure     │ Following BGP peers are not found or routes are not exchanged │ Verifies if BGP peers have correctly advertised/received      │ routing, bgp  │
│           │                          │             │ properly:                                                     │ routes with type as valid and active for a specified VRF.     │               │
│           │                          │             │ {'bgp_peers': {'172.30.11.1': {'default':                     │                                                               │               │
│           │                          │             │ {'advertised_routes': {'192.0.254.31/32': 'Not found'},       │                                                               │               │
│           │                          │             │ 'received_routes': {'192.0.255.4/32': {'valid': False,        │                                                               │               │
│           │                          │             │ 'active': True}}}}, '172.30.11.5': {'default':                │                                                               │               │
│           │                          │             │ {'advertised_routes': {'192.2.254.31/32': 'Not found'},       │                                                               │               │
│           │                          │             │ 'received_routes': {'192.0.255.41/32': 'Not found'}}}}}       │                                                               │               │
└───────────┴──────────────────────────┴─────────────┴───────────────────────────────────────────────────────────────┴───────────────────────────────────────────────────────────────┴───────────────┘
```
Fixes  #482 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have run pre-commit for code linting and typing (`pre-commit run`)
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

**Review by carl-baillargeon — CHANGES_REQUESTED** [review]

File: anta/tests/routing/bgp.py [hunk 1]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
...
+        AntaTemplate(template="show bgp neighbors {neighbor} routes vrf {vrf}"),
+    ]
+
+    class Input(AntaTest.Input):  # pylint: disable=missing-class-docstring
+        bgp_neighbors: List[BgpNeighbors]
+        """List of BGP neighbors"""
+
+        class BgpNeighbors(BaseModel):  # pylint: disable=missing-class-docstring
+            neighbor: Union[IPv4Address, IPv6Address]
+            """IPv4 BGP neighbor"""
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:475]: I did not tested yet, will try once

File: anta/tests/routing/bgp.py [hunk 2]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
...
+            vrf: str = "default"
+            """VRF context"""
+            advertised_routes: list  # type: ignore
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:478]: Fixed

File: anta/tests/routing/bgp.py [hunk 3]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
...
+            """Advertised routes"""
+            received_routes: list  # type: ignore
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:480]: Fixed

File: anta/tests/routing/bgp.py [hunk 4]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
...
+            """Recivied routes"""
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:481]: Fixed

File: anta/tests/routing/bgp.py [hunk 5]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
...
+    def test(self) -> None:
+        failures: dict[dict[str, Any], dict[str, Any]] = {}
+
+        # Itreating over command output for different neighbors
+        for command in self.instance_commands:
+            neighbor = command.params.get("neighbor")
+            vrf = command.params.get("vrf")
+            advertised_routes = command.params.get("advertised_routes")
+            received_routes = command.params.get("received_routes")
+            if not isinstance(advertised_routes, Iterable) or not isinstance(received_routes, Iterable):
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:501]: Updated the code

File: anta/tests/routing/bgp.py [hunk 6]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
...
+                invalid_or_inactive = [
+                    route
+                    for route in received_routes
+                    if route in bgp_routes
+                    and (not bgp_routes[route]["bgpRoutePaths"][0]["routeType"]["valid"] or not bgp_routes[route]["bgpRoutePaths"][0]["routeType"]["active"])
+                ]
+                if missing_routes:
+                    failures = utils.deep_update(failures, {"received_routes": {vrf: {str(neighbor): {"missing_routes": missing_routes}}}})
+                if invalid_or_inactive:
+                    failures = utils.deep_update(failures, {"received_routes": {vrf: {str(neighbor): {"invalid_or_inactive_routes": invalid_or_inactive}}}})
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:541]: created helper function

File: tests/units/anta_tests/routing/test_bgp.py
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply tests/units/anta_tests/routing/test_bgp.py:1]: updated the eos data in unit tests. Thanks for the suggestion

File: anta/tests/routing/bgp.py [hunk 7]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
...
             self.result.is_failure(f"Failures: {list(failures.values())}")
+
+
+class VerifyBGPExchangedRoutes(AntaTest):
+    """
+    Verifies if BGP routes are exchanged for a specified VRF.
+
+    Expected results:
+        * success: The test will pass if BGP routes are exchanged in the specified VRF.
+        * failure: The test will fail if BGP routes are not exchanged in the specified VRF.
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:458]: updated doc string

File: anta/tests/routing/bgp.py [hunk 8]
```diff
@@ -447,3 +447,100 @@ def test(self) -> None:
 
         if failures:
             self.result.is_failure(f"Failures: {list(failures.values())}")
+
+
+class VerifyBGPExchangedRoutes(AntaTest):
+    """
+    Verifies if BGP routes are exchanged for a specified VRF.
```
[the remark this answers is not in the document]
Comment (MaheshGSLAB) [reply anta/tests/routing/bgp.py:454]: updated

File: anta/tests/routing/bgp.py [hunk 9]
```diff
@@ -132,6 +132,61 @@ def _check_peer_issues(peer_data: Optional[dict[str, Any]]) -> dict[str, Any]:
     return {}
 
 
+def _add_bgp_routes_failure(
+    bgp_routes: List[str], bgp_output: dict[str, Any], neighbor: str, vrf: str, route_type: str = "advertised_routes"
```
Comment (gmuloc) [inline anta/tests/routing/bgp.py:136]:
```suggestion
    bgp_routes: list[str], bgp_output: dict[str, Any], neighbor: str, vrf: str, route_type: str = "advertised_routes"
```

File: anta/tests/routing/bgp.py [hunk 10]
```diff
@@ -132,6 +132,61 @@ def _check_peer_issues(peer_data: Optional[dict[str, Any]]) -> dict[str, Any]:
...
+) -> dict[str, dict[str, dict[str, dict[str, list[str]]]]]:
+    """
+    Identifies missing BGP routes and invalid or inactive route entries.
+
+    This function checks the BGP output from the device against the expected routes.
+    It identifies any missing routes as well as any routes that are invalid or inactive. The results are returned in a dictionary.
+
+    Parameters:
+        bgp_routes (List[str]): The list of expected routes.
```
Comment (gmuloc) [inline anta/tests/routing/bgp.py:145]:
```suggestion
        bgp_routes (list[str]): The list of expected routes.
```

File: anta/tests/routing/bgp.py [hunk 11]
```diff
@@ -132,6 +132,61 @@ def _check_peer_issues(peer_data: Optional[dict[str, Any]]) -> dict[str, Any]:
...
+
+    # Prepare the failure routes dictionary
+    failure_routes = {}
+    missing_routes = [route for route in bgp_routes if route not in bgp_output]
+    invalid_or_inactive = [
+        route
+        for route in bgp_routes
+        if route in bgp_output
+        and (not bgp_output[route]["bgpRoutePaths"][0]["routeType"]["valid"] or not bgp_output[route]["bgpRoutePaths"][0]["routeType"]["active"])
+    ]
```
Comment (gmuloc) [inline anta/tests/routing/bgp.py:181]: Here you are looping twice over the `bgp_routes` so it is probably more efficient to use a for loop and loop only once:
```
for route in bgp_routes:
  if not route in bgp_output:
    missing_routes.append(route)
    continue
  if (big condition):
    invalid_or_inactive.append(route)
```

File: anta/tests/routing/bgp.py [hunk 12]
```diff
@@ -447,3 +502,87 @@ def test(self) -> None:
...
+
+        # Iterating over command output for different neighbors
+        for command in self.instance_commands:
+            neighbor = command.params.get("neighbor", "")
+            vrf = command.params.get("vrf", "")
+            advertised_routes = command.params.get("advertised_routes", [])
+            received_routes = command.params.get("received_routes", [])
+
+            # Verify if BGP neighbor is configured with provided vrf
+            if (bgp_routes := get_value(command.json_output, f"vrfs..{vrf}", separator="..")) is None:
```
Comment (gmuloc) [inline anta/tests/routing/bgp.py:567]: Do you need `..` as separator here?

File: anta/tests/routing/bgp.py [hunk 13]
```diff
@@ -447,3 +502,87 @@ def test(self) -> None:
...
+                self.result.is_failure(f"BGP neighbor {str(neighbor)} is not configured for `{vrf}` VRF.")
+                return
```
Comment (gmuloc) [inline anta/tests/routing/bgp.py:569]: if you return here then you won't check all the commands right? So maybe you will miss some failures in other commands that you could report

Would it not be better with
```suggestion
                continue
```
? 

(same below)

File: tests/units/anta_tests/routing/test_bgp.py [hunk 14]
```diff
@@ -1239,4 +1239,216 @@
...
+            ]
+        },
+        "expected": {"result": "success"},
+    },
+    {
+        "name": "failure-no-routes",
+        "test": VerifyBGPExchangedRoutes,
+        "eos_data": [
+            {"vrfs": {"default": {"vrf": "default", "routerId": "192.0.255.1", "asn": "65001", "bgpRouteEntries": {}}}},
+            {"vrfs": {"default": {"vrf": "default", "routerId": "192.0.255.1", "asn": "65001", "bgpRouteEntries": {}}}},
```
Comment (gmuloc) [inline tests/units/anta_tests/routing/test_bgp.py:1300]: so both commands return the same structure everytime? Why do we run both?

File: tests/units/anta_tests/routing/test_bgp.py [hunk 15]
```diff
@@ -1239,4 +1239,216 @@
...
+            },
+        ],
+        "inputs": {
+            "bgp_neighbors": [
+                {
+                    "neighbor": "172.30.11.1",
+                    "vrf": "default",
+                    "advertised_routes": ["192.0.254.31/32"],
+                    "received_routes": ["192.0.255.31/32"],
+                },
```
Comment (gmuloc) [inline tests/units/anta_tests/routing/test_bgp.py:1384]: test with more than one neighbor in your inputs, in particular with one success and one failure and check that the test is failing

**Review by gmuloc — CHANGES_REQUESTED** [review]

File: tests/units/anta_tests/routing/test_bgp.py (hunk 14 above)
Comment (MaheshGSLAB) [reply tests/units/anta_tests/routing/test_bgp.py:1300]: yes both command return same structure output but to check advertised or received routes for a neighbor we need to change the command. From single command we are not getting both advertised and received routes

File: anta/tests/routing/bgp.py [hunk 16]
```diff
@@ -447,3 +493,82 @@ def test(self) -> None:
...
+    @AntaTest.anta_test
+    def test(self) -> None:
+        failures: dict[str, dict[str, Any]] = {}
+
+        # Iterating over command output for different peers
+        for command in self.instance_commands:
+            peer = str(command.params.get("peer", ""))
+            vrf = command.params.get("vrf", "")
+            advertised_routes = command.params.get("advertised_routes", [])
+            received_routes = command.params.get("received_routes", [])
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:555]: As discussed, in recent ANTA refactor, it looks like `command.params` is properly typed as `Dict[str, Any]` so we can simply do a dict lookup without mypy complaining. No need to bother with `.get()` or `cast` anymore.

**Review by carl-baillargeon — CHANGES_REQUESTED** [review]

**Review by gmuloc — APPROVED** [review]

**Pull request merged**